### PR TITLE
Possible fix #24: unwanted attribute folding

### LINKED ISSUE #24: Class names with the same letters as the attribute name will fold the attribute
image
As you can see if you type "c" or "lass", etc, will break the attribute.

I know that this is hard to fix since this is not only used for ``class`` but maybe a fix could be to have 2 regex, one for the attribute and another for the value.

### PR DESCRIPTION
**Pull request by JasperSurmont**
Fixing issue #24 can be done by either [grouping every regex symbol in a group](https://stackoverflow.com/a/1985709/12339252), which could be a better solution than the proposed one here. The downside is that writing the regex for end-users is more complicated and weird.

This solution works by adding a configuration to specify a string that should occur before folding is applied: in the issue's case this would be the character `=` such that no attributes ever get folded `class='class-...'`.

I think this proposed solution is better because people rarely name there classes starting with 'class', hence only a small portion of people need to adjust configuration.

### CODE REVIEW

**Comment by moalamri** [comment]
Hey @JasperSurmont 
Thank you for helping out 🙂 

I'll have a look, I hope it will be better than the solution I came up with.

**Comment by moalamri** [comment]
@JasperSurmont 

That did fix the issue 😀 
Thank you so much.
I might purpose few things tomorrow so it can play well with the upcoming update that supports language scoped settings.

Side question: there's something confuses me as I've see you added the emuns and everything, but didn't add the settings to package.json, so I might guess you hardcoded it first then moved it to enum? Perhaps that's why you had doubts of the solution :D well, 
It worked as expected. Very nice, Thank you so much 👍

**Comment by JasperSurmont** [comment]
Ah true I hardcoded the setting to test and forgot to add it to the package afterwards 😄 

Thanks for the neat plugin!

**Comment by moalamri** [comment]
By the, I had this quick idea, isn't it better to measure the length of the groups before the user specified group? 
Since this extension is a one regex only and users must specify the group to be folded. If so then no need to add extra settings or to limit pre folding to a single character that might only work in single case 🤔.
You brought the solution, I had zero doubts that it was even the code (was blaming regex). Do you think my suggestion could work?
By the way the next update will have two separated decorator options. So there's nothing you're blocking by your solution, it could be handy if I'm going to apply pre and post regex.
If my suggestion wouldn't work then I'll take yours to the next version, perhaps under different name (in the settings) so it can blend with the new language scoped settings (they will be under inlineFold.option...") it could be something like "inlineFold.options.prefold.character" and use it to measure the inset starting point as your solution 👍

**Comment by JasperSurmont** [comment]
Of course, feel free to change anything you want. I just added this to show a possible solution 😄 
As for you suggestion: in the default case it indeed works, but if the user changes the match regex to e.g. `class='([^']+)'`, this would fail as the only group is the one we want to match.

I think the most fail-safe solution is parsing the regex into a new one using the group number: 
`class='([^']+)'` would have to become `(class=')([^']+)'`. In this case we know that the first group's length is always the start index for our to-be-folded group.

I'm not too experienced in regex so correct me if I'm wrong :)

**Comment by moalamri** [comment]
Hey.
So I just had a look, and there's something confuses me.
Your solution works perfectly. Even if any random character is used for the config `matchAfter`, do you experince the same thing?

**Comment by JasperSurmont** [comment]
I don't understand `const skip = match[0].indexOf(matched.replace(match[regexGroup])) + 1;`
Doesn't string.replace() require 2 arguments?

**Comment by moalamri** [comment]
Yes it does need two variables, if I pass empty string it will count, if no replacement is passed then the matched string will be undefined, which is better in this case I guess.

**Comment by JasperSurmont** [comment]
Ah okay I understand. 

Yes that should work and is a better solution than mine :). It would still 'bug' if there are 2 occurences before the folded portion, but I don't think that is something to worry about (the user is definitely doing something wrong then :D)

**Comment by moalamri** [comment]
"Better solution than mine"

Dude I was cursing the regex the whole time. You really saved my day. You brought the solution. Thank you so much for that.

**Comment by JasperSurmont** [comment]
Closing because this solution is not really good and I think I suggested a better one in #24

**Comment by JasperSurmont** [comment]
> @JasperSurmont Please accept the collaboration request to review and merge the PR ;)

Just so you know, you can perform a pull request without making me a contributor, right?

**Comment by JasperSurmont** [comment]
Actually, this solution doesn´t work with { and }, as they are different :thinking:

**Comment by moalamri** [comment]
@JasperSurmont What do you mean they are different? They also measured as 1 character

**Comment by JasperSurmont** [comment]
In: 
```ts
const endsWithChar = fullMatch.charAt(endsWithIndex);
const startWithIndex = fullMatch.indexOf(endsWithChar);
```
The startWithIndex will return the same as endsWithIndex, because there is only 1 `}`, and we are actually looking for `{`

But no worries, I got a solution coming

**Comment by moalamri** [comment]
@JasperSurmont That's a whole rewriting there :D
Well, done. Could you have a look at this and test any desired delimiter, I added a simple method to get the suffix:
https://replit.com/@moalamri/Regex-anatomy#index.ts

File: src/decorator.ts [hunk 1]
```diff
@@ -68,30 +68,43 @@ export class Decorator {
...
+      /* Checking if the range is within the visible area of the editor plus a specified offset for a head decoration. */
       if (!(this.StartLine <= range.start.line && range.end.line <= this.EndLine)) {
         continue;
       }
 
+      /* Pushing the range and the hoverMessage to the decorators array to apply later. */
       decorators.push({
         range,
-        hoverMessage: `Full text **${matched}**`
+        hoverMessage: `Full text **${match[2]}**`
```
Comment (moalamri) [inline src/decorator.ts:107]: This gives wrong full text. match[regexGroup] will get the correct string

**Comment by JasperSurmont** [comment]
Also the parsing of the regex can probably be done on activation of the plugin, not every time the decorators update.

File: src/decorator.ts (hunk 1 above)
Comment (moalamri) [reply src/decorator.ts:107]: Correct `[regexGroup + 1]` no need for extra variable :D

**Comment by moalamri** [comment]
> Also the parsing of the regex can probably be done on activation of the plugin, not every time the decorators update.

Moved the parsed regex to the Decorator scope under the `updateConfigs` method. It will get fired once when the extension is activated and whenever any extension related settings is changed

**Comment by JasperSurmont** [comment]
Seems all good to me!

**Comment by moalamri** [comment]
Well done 🥇 

```log
 [INFO]  Publishing 'moalamri.inline-fold v0.1.6'...
 [INFO]  Extension URL https://marketplace.visualstudio.com/items?itemName=moalamri.inline-fold
 [INFO]  Might take a few minutes
 [INFO]  Hub URL: https://marketplace.visualstudio.com/manage/publishers/moalamri/extensions/inline-fold/hub
 [DONE]  Published moalamri.inline-fold v0.1.6.
```

**Pull request merged**